File: src/chipschallenge/blockreactions/IceskatesTo.java
package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.Block.Type;
import chipschallenge.BlockContainerFullException;
import chipschallenge.Inventory.Boots;

public class IceskatesTo extends BlockReaction {

    private IceskatesTo() {
    }
    private static IceskatesTo mInstance = null;

    public static synchronized IceskatesTo getInstance() {
        if (mInstance == null) {
            mInstance = new IceskatesTo();
        }
        return mInstance;
    }

    public void react(Block moving, Block standing) throws BlockContainerFullException {
        if (moving.isChip()) {
            takeBoots(Boots.ICESKATES);
            standing.replace(createBlock(Type.FLOOR));
        }
    }

    public boolean canMove(Block moving, Block standing) {
        return moving.isChip() || moving.isBlock();
    }
}

File: ChipsChallenge.Shared/Blockreactions/IceskatesTo.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    using Boots = Inventory.Boots;
    using Type = Block.Type;

    public class IceskatesTo : NoSlipReaction
    {
        private static IceskatesTo instance;

        private IceskatesTo()
        {
        }

        public static IceskatesTo Instance
        {
            get
            {
                lock (typeof(IceskatesTo))
                {
                    return instance ?? (instance = new IceskatesTo());
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
            if (moving.Chip)
            {
                TakeBoots(Boots.ICESKATES);
                standing.Replace(CreateBlock(Type.FLOOR));
            }
        }

        public override bool canMove(Block moving, Block standing)
        {
            return moving.Chip || moving.IsBlock();
        }
    }
}